### PR TITLE
Integrated material datepicker and datepicker inputs with CDK datepicker and datepicker inputs

### PR DESCRIPTION
**Pull request by victoriaaa234**
Removed a test that the compiler threw errors on in the `datepicker.spec.ts` in material that I will add back in the CDK in next PR. 

BREAKING CHANGES:
- `MAT_DATE_FORMATS` has been split into `MAT_DATE_FORMATS` and `CDK_DATE_FORMATS`. In most cases this will not effect existing apps, as components that use `CDK_DATE_FORMATS` currentlly fall back to looking at `MAT_DATE_FORMATS` if it is not found. However, if you are overriding `MAT_DATE_FORMATS` for some portion of your app, you must now also override `CDK_DATE_FORMATS`. A quick fix is to override it to undefined, thus forcing components to fall back to `MAT_DATE_FORMATS` again. (e.g. `{provide: CDK_DATE_FORMATS, useValue: undefined}`). Ideally you should split out a `CDK_DATE_FORMATS` value from your existing `MAT_DATE_FORMATS` value however, as this fallback behavior will be removed in the future.

### CODE REVIEW

File: src/cdk/datetime/datepicker-input.ts [hunk 1]
```diff
@@ -197,33 +197,36 @@ export class CdkDatepickerInput<D> implements AfterContentInit, ControlValueAcce
   /** Implemented for datepicker CDK and locale subscriptions. */
   private readonly _subscriptions = new Subscription();
 
+  /** Prefix for form control validator properties. */
+  protected _prefix = 'cdk';
```
Comment (jelbourn) [inline src/cdk/datetime/datepicker-input.ts]: The name `_prefix` is a bit generic- can you make this more specific?

File: src/lib/datepicker/datepicker-input.ts [hunk 2]
```diff
@@ -79,6 +86,8 @@ export class MatDatepickerInputEvent<D> {
     MAT_DATEPICKER_VALIDATORS,
     {provide: MAT_INPUT_VALUE_ACCESSOR, useExisting: MatDatepickerInput},
   ],
+  inputs: ['value', 'min', 'max', 'disabled', 'cdkDatepickerFilter', 'cdkDatepicker'],
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker-input.ts]: I think we have to rename `'cdkDatepickerFilter'` and `'cdkDatepicker'`

File: src/lib/datepicker/datepicker-input.ts [hunk 3]
```diff
@@ -92,198 +101,35 @@ export class MatDatepickerInputEvent<D> {
...
+export class MatDatepickerInput<D> extends CdkDatepickerInput<D> implements AfterContentInit,
+    ControlValueAccessor, OnDestroy, Validator {
+  /** Prefix for form control validator properties. */
+  protected _formControlValidatorPrefix = 'mat';
+
   /** The datepicker that this input is associated with. */
   @Input()
   set matDatepicker(value: MatDatepicker<D>) {
-    this.registerDatepicker(value);
+    this.datepicker = value;
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker-input.ts]: I _think_ we can just rename these in the `inputs` list above

File: src/lib/datepicker/datepicker.spec.ts [hunk 4]
```diff
@@ -1379,13 +1379,6 @@ describe('MatDatepicker', () => {
 
   });
 
-  describe('with missing DateAdapter and MAT_DATE_FORMATS', () => {
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker.spec.ts]: Why was this removed? I would think it would still throw

File: src/lib/datepicker/datepicker.ts [hunk 5]
```diff
@@ -253,51 +201,29 @@ export class MatDatepicker<D> implements OnDestroy, CanColor {
...
-  /** Emits new selected date when selected date changes. */
-  readonly _selectedChanged = new Subject<D>();
-
   constructor(private _dialog: MatDialog,
               private _overlay: Overlay,
               private _ngZone: NgZone,
               private _viewContainerRef: ViewContainerRef,
               @Inject(MAT_DATEPICKER_SCROLL_STRATEGY) private _scrollStrategy,
-              @Optional() private _dateAdapter: DateAdapter<D>,
+              @Optional() protected _dateAdapter: DateAdapter<D>,
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker.ts]: remove `protected`, super class will save a reference for you

File: src/cdk/datetime/datepicker-input.ts [hunk 6]
```diff
@@ -237,7 +245,7 @@ export class CdkDatepickerInput<D> implements AfterContentInit, ControlValueAcce
   constructor(
       protected _elementRef: ElementRef,
       @Optional() public _dateAdapter: DateAdapter<D>,
-      @Optional() @Inject(CDK_DATE_FORMATS) private _dateFormats: CdkDateFormats) {
+      @Optional() @Inject(CDK_DATE_FORMATS) protected _dateFormats: CdkDateFormats) {
```
Comment (mmalerba) [inline src/cdk/datetime/datepicker-input.ts]: Grab me offline to discuss `CDK_DATE_FORMATS` and `MAT_DATE_FORMATS`

File: src/lib/datepicker/datepicker-input.ts (hunk 3 above)
Comment (victoriaaa234) [reply src/lib/datepicker/datepicker-input.ts]: Renaming in the `inputs` above didn't work, maybe setter/getter problem? We discussed this and decided to leave this as is.

File: src/lib/datepicker/datepicker.spec.ts (hunk 4 above)
Comment (victoriaaa234) [reply src/lib/datepicker/datepicker.spec.ts]: I put it back but referred to the `CdkDatepicker` in the test rather than the `MatDatepicker`.

File: src/cdk/datetime/datepicker-input.ts (hunk 6 above)
Comment (victoriaaa234) [reply src/cdk/datetime/datepicker-input.ts]: Going to separate `CDK_DATE_FORMATS` entirely from `MAT_DATE_FORMATS`.

File: src/lib/datepicker/datepicker-input.ts (hunk 2 above)
Comment (victoriaaa234) [reply src/lib/datepicker/datepicker-input.ts]: Already removed them :)

File: src/lib/datepicker/datepicker-input.ts [hunk 7]
```diff
@@ -92,198 +101,35 @@ export class MatDatepickerInputEvent<D> {
...
-
-  /** @docs-private */
-  registerOnValidatorChange(fn: () => void): void {
-    this._validatorOnChange = fn;
-  }
-
-  /** @docs-private */
-  validate(c: AbstractControl): ValidationErrors | null {
-    return this._validator ? this._validator(c) : null;
+    super(_elementRef, _dateAdapter, _dateFormats);
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker-input.ts]: in order to keep it non-breaking, inject `MAT_DATE_FORMATS` and `CDK_DATE_FORMATS`, then pass `_cdkDateFormats || _matDateFormats` to the constructor.

Mark the `MAT_DATE_FORMATS` param as deprecated

File: src/material-moment-adapter/adapter/moment-date-formats.ts [hunk 8]
```diff
@@ -7,16 +7,34 @@
...
+     * @deletion-target 8.0.0
+     */
     dateInput: 'l',
     monthYearLabel: 'MMM YYYY',
     dateA11yLabel: 'LL',
     monthYearA11yLabel: 'MMMM YYYY',
   },
 };
+
+export const CDK_MOMENT_DATE_FORMATS: CdkDateFormats = {
```
Comment (mmalerba) [inline src/material-moment-adapter/adapter/moment-date-formats.ts]: you need to provide this in the moment module as well

Comment (victoriaaa234) [reply src/material-moment-adapter/adapter/moment-date-formats.ts]: Fixed changes

File: src/lib/datepicker/datepicker-input.ts [hunk 9]
```diff
@@ -122,9 +126,14 @@ export class MatDatepickerInput<D> extends CdkDatepickerInput<D> implements Afte
   constructor(
       _elementRef: ElementRef,
       @Optional() _dateAdapter: DateAdapter<D>,
-      @Optional() @Inject(CDK_DATE_FORMATS) _dateFormats: CdkDateFormats,
+      @Optional() @Inject(CDK_DATE_FORMATS) _cdkDateFormats: CdkDateFormats,
```
Comment (mmalerba) [inline src/lib/datepicker/datepicker-input.ts]: move this param to the end and make it optional (`_cdkDateFormats?`) also put a comment `/** @deletion-target 8.0.0 Make required */`
